### PR TITLE
Add payments canceling by super major disapprove #1105

### PR DESCRIPTION
**Pull request by maslenitsa93**
#1105

### CODE REVIEW

File: libraries/chain/worker_evaluators.cpp [hunk 1]
```diff
@@ -323,7 +323,7 @@ namespace golos { namespace chain {
             logic_exception::approver_of_result_should_be_in_top19_of_witnesses,
             "Approver of result should be in Top 19 of witnesses");
 
-        const auto& wto_post = _db.get_comment(o.author, o.permlink);
+        const auto& wro_post = _db.get_comment(o.author, o.permlink);
         const auto& wto = _db.get_worker_result(o.author, o.permlink);
```
Comment (afalaleev) [inline libraries/chain/worker_evaluators.cpp]: Why `wto` instead of `wro`?

File: libraries/chain/worker_evaluators.cpp [hunk 2]
```diff
@@ -480,4 +468,75 @@ namespace golos { namespace chain {
...
+        GOLOS_CHECK_LOGIC(approver_witness.schedule == witness_object::top19,
+            logic_exception::approver_of_payment_should_be_in_top19_of_witnesses,
+            "Approver of payment should be in Top 19 of witnesses");
+
+        GOLOS_CHECK_LOGIC(o.state != worker_techspec_approve_state::approve,
+            logic_exception::payment_does_not_need_approving_it_can_be_only_disapproved,
+            "Payment does not need approving, it can be only disapproved");
+
+        const auto& wro_post = _db.get_comment(o.worker_result_author, o.worker_result_permlink);
+        const auto& wto = _db.get_worker_result(o.worker_result_author, o.worker_result_permlink);
```
Comment (afalaleev) [inline libraries/chain/worker_evaluators.cpp]: Why `wto`?

File: libraries/chain/worker_evaluators.cpp [hunk 3]
```diff
@@ -480,4 +468,75 @@ namespace golos { namespace chain {
...
+
+        const auto& wpo_post = _db.get_comment(wto.worker_proposal_author, wto.worker_proposal_permlink);
+        const auto& wpo = _db.get_worker_proposal(wpo_post.id);
+
+        GOLOS_CHECK_LOGIC(wpo.state == worker_proposal_state::payment,
```
Comment (afalaleev) [inline libraries/chain/worker_evaluators.cpp]: It seems strange, that proposal has payment state.
The tspec is received payments, so proposal should has only state - tspec, all other are states of tspec.

Comment (afalaleev) [reply libraries/chain/worker_evaluators.cpp]: Lets fix it in another issue, not here.

File: libraries/chain/worker_evaluators.cpp [hunk 4]
```diff
@@ -480,4 +468,75 @@ namespace golos { namespace chain {
...
+                wpo.state = worker_proposal_state::created;
+                return;
+            }
+
+            wpo.state = worker_proposal_state::work;
+        });
+
+        _db.modify(wto, [&](worker_techspec_object& wto) {
+            wto.next_cashout_time = time_point_sec::maximum();
+            wto.month_consumption = asset(0, STEEM_SYMBOL);
```
Comment (afalaleev) [inline libraries/chain/worker_evaluators.cpp]: Lets add issue to extend `worker_techspec_state` with extra states.

File: libraries/chain/worker_evaluators.cpp (hunk 3 above)
Comment (maslenitsa93) [reply libraries/chain/worker_evaluators.cpp]: @afalaleev Fixed https://github.com/GolosChain/golos/pull/1137

File: libraries/chain/worker_evaluators.cpp (hunk 4 above)
Comment (maslenitsa93) [reply libraries/chain/worker_evaluators.cpp]: @afalaleev Fixed https://github.com/GolosChain/golos/pull/1137

File: libraries/chain/worker_evaluators.cpp [hunk 5]
```diff
@@ -441,4 +439,78 @@ namespace golos { namespace chain {
...
+        ASSERT_REQ_HF(STEEMIT_HARDFORK_0_21__1013, "worker_payment_approve_operation");
+
+        auto approver_witness = _db.get_witness(o.approver);
+        GOLOS_CHECK_LOGIC(approver_witness.schedule == witness_object::top19,
+            logic_exception::approver_of_payment_should_be_in_top19_of_witnesses,
+            "Approver of payment should be in Top 19 of witnesses");
+
+        GOLOS_CHECK_LOGIC(o.state != worker_techspec_approve_state::approve,
+            logic_exception::payment_does_not_need_approving_it_can_be_only_disapproved,
+            "Payment does not need approving, it can be only disapproved");
```
Comment (afalaleev) [inline libraries/chain/worker_evaluators.cpp]: Example.
I'm witness.
I see payments for existing work.
I disapprove it.
But some time later I understand that work is really good, and I was mistaken.
I can approve work, and chain should remove my vote from the state.
Why can't I approve work?

File: libraries/chain/worker_evaluators.cpp [hunk 6]
```diff
@@ -470,17 +466,15 @@ namespace golos { namespace chain {
...
-            _db.modify(*wpao_itr, [&](worker_payment_approve_object& wpao) {
-                wpao.state = o.state;
-            });
-        } else {
-            _db.create<worker_payment_approve_object>([&](worker_payment_approve_object& wpao) {
-                wpao.approver = o.approver;
-                wpao.post = wro_post.id;
-                wpao.state = o.state;
-            });
+            return;
```
Comment (afalaleev) [inline libraries/chain/worker_evaluators.cpp]: Here should be error - operation shouldn't be included into block.

Comment (maslenitsa93) [reply libraries/chain/worker_evaluators.cpp]: @afalaleev https://github.com/GolosChain/golos/issues/1139

File: libraries/chain/worker_evaluators.cpp [hunk 7]
```diff
@@ -441,4 +439,68 @@ namespace golos { namespace chain {
...
+        }
+
+        const auto& gpo = _db.get_dynamic_global_properties();
+        _db.modify(gpo, [&](dynamic_global_property_object& gpo) {
+            gpo.worker_consumption_per_month -= _db.calculate_worker_techspec_month_consumption(wto);
+        });
+
+        const auto& wpo_post = _db.get_comment(wto.worker_proposal_author, wto.worker_proposal_permlink);
+        const auto& wpo = _db.get_worker_proposal(wpo_post.id);
+        if (wpo.type == worker_proposal_type::premade_work) {
```
Comment (afalaleev) [inline libraries/chain/worker_evaluators.cpp]: For what this checking?
You do same things in this if-block and after it.

**Comment by afalaleev** [comment]
Fix conflicts

**Pull request merged**